### PR TITLE
Reduce the frequency of job status update and remove parallel query

### LINKED ISSUE #1088: [Spot] Ray dashboard hangs making `ray job` commands not responding
Although not much CPU and memory are used, the `sky-spot-controller` can still fail to take new `ray job` commands, due to a lot of `ray job` commands running and the jobs are stuck in INIT mode. That causes the `sky spot launch` to hang after setup is completed.

One possible reason for that is our skylet will try to update the status of the INIT jobs by querying ray with `ray job`, causing a lot of connection to the ray dashboard, `http://127.0.0.1:8265`.

It happens in the situation `In progress jobs: 14 PENDING, 4 RUNNING, 4 STARTING`, but I failed to reproduce the problem when `In progress jobs: 29 PENDING, 8 RUNNING` with the following code:
import subprocess

```
for i in range(200):
    for j in range(5):
        # retry
        try:
            subprocess.run(f'sky spot launch -y -d -n test-{i} job.yaml', shell=True, check=True)
        except subprocess.CalledProcessError:
            continue
        break
```

```
# job.yaml
resources:
  cloud: gcp
  region: us-west1

run: echo hi; sleep 800; echo bye

```

### PR DESCRIPTION
**Pull request by Michaelvll**
This is an attempt for #1088. It reduces the parallel query through the same port for the ray job.

Profiling: (The original time is not that large (reduced by 3x), but the list_jobs can reduce the number of http requests by the number of jobs to be queried.)
```bash
$ python ./query_jobs.py 
Parallel Time spend: 0.10391879081726074
job list Time spend: 0.03846001625061035
```
```python
# query_jobs.py
# 10 jobs not exist, 11 RUNNING, 3 STARTING, 6 SUCCEEDED
from sky.skylet import job_lib
from sky.utils import subprocess_utils
import time

client = job_lib._create_ray_job_submission_client()

ray_job_ids = [f'{i}-gcpuser' for i in range(60, 90)]
def get_job_status(job_id):
    try:
        # The return value is a string, e.g. 'RUNNING', which conflicts
        # with the return type in ray code.
        ray_status = client.get_job_status(job_id)
        return job_lib._RAY_TO_JOB_STATUS_MAP[ray_status]
    except RuntimeError as e:
        # If the job does not exist or if the request to the
        # job server fails.
        if 'does not exist' in str(e):
            return None
        raise

start = time.time()
ray_statuses = subprocess_utils.run_in_parallel(
    get_job_status, ray_job_ids)
print(f'Parallel Time spend: {time.time() - start}')

client = job_lib._create_ray_job_submission_client()
start = time.time()
ray_job_infos = client.list_jobs()
job_statuses = [None] * len(ray_job_ids)
for i, ray_job_id in enumerate(ray_job_ids):
    if ray_job_id in ray_job_infos:
        ray_status = ray_job_infos[ray_job_id].status
        job_statuses[i] = job_lib._RAY_TO_JOB_STATUS_MAP[ray_status]
print(f'job list Time spend: {time.time() - start}')
```

Tested:
- [x] `pytest ./tests/test_smoke.py::test_stale_job`
- [x] `pytest ./tests/test_smoke.py::test_stale_job_manual_restart`

### CODE REVIEW

File: sky/skylet/job_lib.py [hunk 1]
```diff
@@ -341,21 +341,15 @@ def update_job_status(job_owner: str,
...
-        except RuntimeError as e:
-            # If the job does not exist or if the request to the
-            # job server fails.
-            if 'does not exist' in str(e):
-                return None
-            raise
+    ray_job_infos = job_client.list_jobs()
+    ray_statuses: List[JobStatus] = []
+    for ray_job_id in ray_job_ids:
+        if ray_job_id not in ray_job_infos:
```
Comment (concretevitamin) [inline sky/skylet/job_lib.py:347]: Interestingly, in Ray master branch `ray_job_infos` becomes a List[JobDetails]

https://github.com/ray-project/ray/blob/master/dashboard/modules/job/sdk.py#L225

Perhaps note in a comment it's a Dict of job id -> JobDetails as of 1.13.0.

File: sky/skylet/job_lib.py [hunk 2]
```diff
@@ -341,21 +341,15 @@ def update_job_status(job_owner: str,
...
-            # with the return type in ray code.
-            ray_status = job_client.get_job_status(job_id)
-            return _RAY_TO_JOB_STATUS_MAP[ray_status]
-        except RuntimeError as e:
-            # If the job does not exist or if the request to the
-            # job server fails.
-            if 'does not exist' in str(e):
-                return None
-            raise
+    ray_job_infos = job_client.list_jobs()
```
Comment (concretevitamin) [inline sky/skylet/job_lib.py:344]: Do you have a sense of the speedup due to this PR, perhaps add to PR description?

I was assuming the max concurrent jobs is single digit, but there can be lots of INIT jobs?

https://github.com/skypilot-org/skypilot/blob/5addab7b67428a6c2d02132d3c2672da4576c451/sky/skylet/job_lib.py#L420

Suggestion: update this variable to say `non_terminal_jobs`; was confused a bit by "running" (which would be single digit and thus doesn't need this change?).

File: sky/skylet/job_lib.py [hunk 3]
```diff
@@ -341,21 +341,15 @@ def update_job_status(job_owner: str,
...
+    ray_statuses: List[JobStatus] = []
+    for ray_job_id in ray_job_ids:
+        if ray_job_id not in ray_job_infos:
+            ray_statuses.append(None)
+            continue
+        ray_status = ray_job_infos[ray_job_id].status
+        ray_statuses.append(_RAY_TO_JOB_STATUS_MAP[ray_status])
```
Comment (concretevitamin) [inline sky/skylet/job_lib.py:351]: If this is a confirmed hot spot, we can do
```
ray_statuses = [None] * len(ray_job_ids)
...
ray_statuses[i] = ...
```
Which reminds me, should we rename `ray_statuses` to `job_statuses`, as it is our semantics?

File: sky/spot/spot_utils.py [hunk 4]
```diff
@@ -116,7 +116,7 @@ def cancel_jobs_by_id(job_ids: Optional[List[int]]) -> str:
         controller_status = job_lib.get_status(job_id)
         if controller_status.is_terminal():
             logger.error(f'Controller for job {job_id} have exited abnormally. '
-                         'Set the job status to FAILED.')
+                         'Set the job status to FAILED_CONTROLLER.')
```
Comment (concretevitamin) [inline sky/spot/spot_utils.py:119]: nit: Set -> Setting

File: tests/test_smoke.py [hunk 5]
```diff
@@ -187,6 +188,34 @@ def test_stale_job():
...
+            f'Name=tag:ray-cluster-name,Values={name} '
+            f'--query Reservations[].Instances[].InstanceId '
+            '--output text`; '
+            f'aws ec2 stop-instances --region {region} '
+            '--instance-ids $id',
+            'sleep 40',
+            f'sky launch -c {name} -y "echo hi"',
+            f'sky logs {name} 1 --status',
+            f'sky logs {name} 3 --status',
+            # Ensure the skylet updated the staled job status.
```
Comment (concretevitamin) [inline tests/test_smoke.py:210]: nit: staled -> stale

**Review by concretevitamin — APPROVED** [review]
Nice @Michaelvll!

File: sky/skylet/job_lib.py (hunk 3 above)
Comment (Michaelvll) [reply sky/skylet/job_lib.py]: Good point! Fixed. Thanks!

File: sky/skylet/job_lib.py (hunk 1 above)
Comment (Michaelvll) [reply sky/skylet/job_lib.py:349]: Good point! Fixed. Thanks!

File: sky/skylet/job_lib.py (hunk 2 above)
Comment (Michaelvll) [reply sky/skylet/job_lib.py:346]: Good point! Renamed the variable.

For the speedup, I added it in the description. The original time is not that large (reduced by 3x), but the list_jobs can reduce the number of http requests by the number of jobs to be queried.

**Comment by Michaelvll** [comment]
Tested again with the two tests, merging for reducing the HTTP requests.
- [x] `pytest ./tests/test_smoke.py::test_stale_job`
- [x] `pytest ./tests/test_smoke.py::test_stale_job_manual_restart`

**Pull request merged**